### PR TITLE
RSA: Refactor `rsa::KeyPair::sign()` in preparation for adding RSA encryption

### PR DESCRIPTION
**Pull request by briansmith**
Refactor the implementation of `rsa::KeyPair::sign()` so that the arithmetic can be reused for RSA encryption. Clarify the implementation design decisions.

Also clarify the API, especially w.r.t. failure conditions.

### CODE REVIEW

File: src/rsa/keypair.rs [hunk 1]
```diff
@@ -563,21 +569,48 @@ impl KeyPair {
...
+        // reduce the required stack space.
+        padding_alg.encode(m_hash, signature, self.public().n().len_bits(), rng)?;
+
+        // RFC 8017 Section 5.1.2: RSADP, using the Chinese Remainder Theorem
+        // with Garner's algorithm.
+
+        // Steps 1 and 2.
+        let m = self.private_exponentiate(signature)?;
+
+        // Step 3.
```
Comment (complexspaces) [inline src/rsa/keypair.rs:588]: Should this still keep the reference to "Manger's Attack revisited"?

File: tests/rsa_tests.rs [hunk 2]
```diff
@@ -142,24 +142,19 @@ fn test_signature_rsa_pkcs1_sign_output_buffer_len() {
...
-    assert!(key_pair
-        .sign(&signature::RSA_PKCS1_SHA256, &rng, MESSAGE, &mut signature)
-        .is_ok());
-
-    // The output buffer is one byte too long.
-    signature.push(0);
-    assert!(key_pair
-        .sign(&signature::RSA_PKCS1_SHA256, &rng, MESSAGE, &mut signature)
-        .is_err());
+    // Wen the output buffer is not exactly the right length, `sign()` returns
```
Comment (complexspaces) [inline tests/rsa_tests.rs:145]: Typo: 
```suggestion
    // When the output buffer is not exactly the right length, `sign()` returns
```

File: tests/rsa_tests.rs [hunk 3]
```diff
@@ -142,24 +142,19 @@ fn test_signature_rsa_pkcs1_sign_output_buffer_len() {
...
+    // an error (and does not panic or invoke UB). if `sign` doesn't check that
```
Comment (complexspaces) [inline tests/rsa_tests.rs:146]: Based on this comment, shouldn't there be some tests checking the error conditions like before?

Comment (briansmith) [reply tests/rsa_tests.rs:146]: Sorry, I don't understand what you mean. Note that the test is testing that the condition "the buffer is the right length" is equal to (`assert_eq!`) the condition "the result is Ok(_).".

File: src/rsa/keypair.rs (hunk 1 above)
Comment (briansmith) [reply src/rsa/keypair.rs:588]: I mentioned why I removed the reference in the third commit message.

File: tests/rsa_tests.rs (hunk 2 above)
Comment (briansmith) [reply tests/rsa_tests.rs]: Thanks. I fixed it and pushed a new version of this set of commits.

File: src/rsa/keypair.rs (hunk 1 above)
Comment (complexspaces) [reply src/rsa/keypair.rs:588]: Ah ok, thank you for the clarification.

File: tests/rsa_tests.rs (hunk 3 above)
Comment (complexspaces) [reply tests/rsa_tests.rs:146]: Oh, this is my bad. I re-read the test and noticed that the left-hand side of the `assert_eq!()` is dependent on `len`'s value. I missed that on my first read-through and combined with the comment, was tripped up.

**Review by complexspaces — APPROVED** [review]

**Comment by briansmith** [comment]
I had to re-push this because rustfmt.

**Pull request merged**